### PR TITLE
Non-recursive evaluator

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: pil-analyzer/src/evaluator.rs [hunk 1]
```diff
@@ -551,51 +516,219 @@ pub trait SymbolLookup<'a, T> {
...
+                    .push(evaluate_literal(n.clone(), ty, &self.type_args)?)
+            }
+            Expression::String(s) => self.value_stack.push(Value::String(s.clone()).into()),
+            Expression::Tuple(_)
+            | Expression::ArrayLiteral(_)
+            | Expression::BinaryOperation(_, _, _)
+            | Expression::UnaryOperation(_, _)
+            | Expression::IndexAccess(_)
+            | Expression::FunctionCall(_) => {
+                let children_count = expr.children().count();
```
Comment (chriseth) [inline pil-analyzer/src/evaluator.rs:641]: While it looks neat to handle all those cases in a uniform way, it might be quite bad for performance.

Comment (chriseth) [reply pil-analyzer/src/evaluator.rs:645]: Indeed it is.

File: Cargo.toml [hunk 2]
```diff
@@ -45,3 +45,7 @@ overflow-checks = true
 panic = 'unwind'
 incremental = true         # This is true because target is cached
 codegen-units = 256
+
+[profile.release-with-debug]
```
Comment (chriseth) [inline Cargo.toml:49]: Will remove that again, or maybe we can also just keep it?

File: pil-analyzer/src/evaluator.rs [hunk 3]
```diff
@@ -551,51 +516,248 @@ pub trait SymbolLookup<'a, T> {
...
+                    let result = self.value_stack.pop().unwrap();
+                    self.symbols.add_constraints(result, SourceRef::unknown())?;
+                }
+            };
+        }
+        assert_eq!(self.value_stack.len(), 1);
+        Ok(self.value_stack.pop().unwrap())
+    }
+
+    /// Evaluate a leaf expression or expand a complex expression.
```
Comment (leonardoalt) [inline pil-analyzer/src/evaluator.rs:624]:
```suggestion
    /// Evaluates a leaf expression or expands a complex expression.
```

File: pil-analyzer/src/evaluator.rs [hunk 4]
```diff
@@ -551,51 +516,248 @@ pub trait SymbolLookup<'a, T> {
...
+    /// The second field is the number of children.
+    Combine(&'a Expression, usize),
+    /// Truncate the local variables to a given length
+    TruncateLocals(usize),
+    /// Replace the environment (local variables and type args).
+    SetEnvironment(Vec<Arc<Value<'a, T>>>, HashMap<String, Type>),
+    /// Evaluate a let statement, adding matched pattern variables to the local variables.
+    LetStatement(&'a LetStatementInsideBlock<Expression>),
+    /// Add a constraint to the constraint set.
+    AddConstraint(),
```
Comment (leonardoalt) [inline pil-analyzer/src/evaluator.rs:534]:
```suggestion
    AddConstraint,
```

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**